### PR TITLE
Create namespace 'scheduler' and class scheduler::component

### PR DESCRIPTION
**Pull request by clemahieu**
This component encloses the buckets/hinted/optimistic schedulers. There are an increasing number of classes that are associated with election scheduling so there is a desire to put these related concepts in their own namespace and file folder.

Move optimistic_scheduler under scheduler/optimistic scheduler::optimistic node.scheduler.optimistic
Move hinted_scheduler under scheduler/hinted scheduler::hinted node.scheduler.hinted
Move election_scheduler under scheduler/component scheduler::component node.scheduler.buckets
Created scheduler::component which contains each of the separate schedulers as one component

Node and components use PIML pattern with convenience references.

Fixed circular reference between active_transactions and scheduler::buckets by using the nano::node & already inside active_transactions.

### CODE REVIEW

File: nano/node/scheduler/buckets.hpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#include <condition_variable>
+#include <deque>
+#include <memory>
+#include <thread>
+
+namespace nano
+{
+class block;
+class node;
+namespace scheduler
```
Comment (pwojcikdev) [inline nano/node/scheduler/buckets.hpp]: Maybe consider specifying the `nano::scheduler` namespace in a single statement (with forward declarations in separate block). Nesting namespaces introduces additional indentation, we do both in our codebase which is inconsistent. Using non-nested version makes it consistent with the most common case.

File: nano/node/scheduler/component.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+	class optimistic;
+
+	class component
+	{
+		std::unique_ptr<nano::scheduler::optimistic> optimistic_impl;
+		std::unique_ptr<nano::scheduler::buckets> buckets_impl;
+		std::unique_ptr<nano::scheduler::hinted> hinted_impl;
+
+	public:
+		component (nano::node & node);
```
Comment (pwojcikdev) [inline nano/node/scheduler/component.hpp]: Code analyzer is going to complain about missing explicit

File: nano/node/scheduler/hinted.hpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+class node;
+class node_config;
+class active_transactions;
+class vote_cache;
+class online_reps;
+
+namespace scheduler
+{
+	/*
+ * Monitors inactive vote cache and schedules elections with the highest observed vote tally.
```
Comment (pwojcikdev) [inline nano/node/scheduler/hinted.hpp:21]: Nesting namespaces introduces problems with inconsistent indentation exactly like visible in this file.

**Review by pwojcikdev — DISMISSED** [review]
It's OK. However, I don't think "Fixed circular reference between active_transactions and scheduler::buckets by using the nano::node & already inside active_transactions." is entirely accurate, since circular dependency is still there, it's just hidden by node which acts as service locator. One could argue that this pattern is even worse. The dependency is only inside `nano::active_transactions::block_cemented_callback` which definitely does too many things and can be split further to properly remove the ugly interdependencies (similar refactoring to what was done for block processor recently).

**Comment by clemahieu** [comment]
Incorporated non-nested namespaces and converted free-function to constructor for hinted::config

**Review by pwojcikdev — APPROVED** [review]

**Pull request merged**